vllm/v1/sample/rejection_sampler.py

```diff
@@ -131,6 +141,123 @@ def parse_output(
+                                      repeat_indices, output_token_ids)
+        logits = self.apply_bad_words(logits, sampling_metadata, metadata,
+                                      output_token_ids)
+        logits = self.apply_allowed_token_ids(logits, sampling_metadata,
+                                              repeat_indices)
+
+        return logits
+
+    def apply_penalties(
+        self,
+        logits: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+        metadata: SpecDecodeMetadata,
+        repeat_indices: Optional[torch.Tensor],
+        output_token_ids: list[list[int]],
+    ) -> torch.Tensor:
+        if sampling_metadata.no_penalties:
+            return logits
+
+        assert sampling_metadata.prompt_token_ids is not None
+
+        prompt_token_ids = sampling_metadata.prompt_token_ids[repeat_indices]
+        presence_penalties = sampling_metadata.presence_penalties[
+            repeat_indices]
+        frequency_penalties = sampling_metadata.frequency_penalties[
+            repeat_indices]
+        repetition_penalties = sampling_metadata.repetition_penalties[
+            repeat_indices]
+
+        logits = apply_all_penalties(
+            logits,
+            prompt_token_ids,
+            presence_penalties,
+            frequency_penalties,
+            repetition_penalties,
+            output_token_ids,
+        )
+        return logits
+
+    def _combine_outputs_with_spec_tokens(
+        self,
+        output_token_ids: list[list[int]],
+        spec_token_ids: Optional[list[list[int]]] = None,
+    ):
+        if spec_token_ids is None:
+            return output_token_ids
+
+        result = []
+        for out, spec in zip(output_token_ids, spec_token_ids):
+            if len(spec) == 0:
+                continue
+            result.append(out)
+            for i in range(len(spec) - 1):
+                result.append([*result[-1], spec[i]])
+        return result
+
+    def apply_bad_words(
+        self,
+        logits: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+        metadata: SpecDecodeMetadata,
+        output_token_ids: list[list[int]],
+    ) -> torch.Tensor:
+        if sampling_metadata.bad_words_token_ids:
+
+            apply_bad_words_with_drafts(
+                logits,
+                sampling_metadata.bad_words_token_ids,
+                output_token_ids,
+                metadata.num_draft_tokens,
+            )
+        return logits
+
+    def apply_allowed_token_ids(
+        self,
+        logits: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+        repeat_indices: torch.Tensor,
+    ) -> torch.Tensor:
+        if sampling_metadata.allowed_token_ids_mask is not None:
+            allowed_token_ids_mask = sampling_metadata.allowed_token_ids_mask[
+                repeat_indices]
+            logits.masked_fill_(allowed_token_ids_mask, float("-inf"))
+
+        return logits
```

Comment:
probably simpler to just inline these two in `apply_logits_processors` ... same for the ones in sampler.py